addons/stock/report/report_stock_rule.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models, _
+from odoo.exceptions import UserError
+
+
+class ReportStockRule(models.AbstractModel):
+    _name = 'report.stock.report_stock_rule'
+
+    @api.model
+    def _get_report_values(self, docids, data=None):
+        product = self.env['product.product'].browse(data['product_id'])
+        warehouses = self.env['stock.warehouse'].browse(data['warehouse_ids'])
+
+        routes = self._get_routes(data)
+
+        # Some routes don't have a warehouse_id but contain rules of diferent warehouses
+        rules = routes.mapped('rule_ids').filtered(lambda r: not r.warehouse_id or r.warehouse_id in warehouses)
+        rules_and_loc = []
+        for rule in rules:
+            rules_and_loc.append(self._get_rule_loc(rule, product))
+
```

Comment:
Here we handle reordering rules and putaway strategies by creating the header_lines dict. This dict is indexed by location_id and contains itself another dict with the relevant reordering rules and putaway strategies.